### PR TITLE
 Restore the mergeTermFreqNormLocsByCopying optimisation

### PR DESCRIPTION
**Pull request by sreekanth-cb**
Even with custom or different chunk size/factors, it still
 looks safe and efficient to copy the freq,norm,locs bytes
 for saving the extra encodings of many uvarint's.

 Chunk factor differences across postings per segment
 doesn't look to affect the functionality here.

 By restoring this optimization, the performance of the merge
 paths improved.  QPS improvements are,

    wildcard - 30%
    fuzzy2 - 34%

 for showfast tests.
 ref - https://issues.couchbase.com/browse/MB-41399

### CODE REVIEW

**Comment by sreekanth-cb** [comment]
Current master during persist loop's merge ops,
Screenshot 2020-11-02 at 5 44 41 PM


On toy with optimization restored, (during persist loop's merge ops)
Screenshot 2020-11-02 at 5 45 21 PM

**Comment by mschoch** [comment]
OK, so going  back to why I thought this wasn't possible.  I think perhaps I misunderstood the original "copying" approach.  I had assumed we were simply copying entire runs of data.  As in, we would copy an entire freq-locs section for a field, and then do that again for each segment.  Basically, just placing them side by side, and my thinking was that this would be incorrect if the chunk factors changed.

However, looking closer at the implementation, it seems that isn't what happens.  We still "rechunk", the difference is that the freq/loc information (per doc) is copied, as opposed to re-encoded.  @sreekanth-cb can you confirm this is the correct understanding?

Second question, does this fix give any understanding as to why we only observe the problem on a single (very repeatably)?

marty

**Comment by sreekanth-cb** [comment]
- Yes, understanding is the same as I have here too.

- Not really. (copying my thoughts/findings from the MB)

_Two primary slowness factors identified/suspected so far for slow merge ops are._

_1.  With the new custom chunk factor introduction in zap, we had to figure out the cardinality of each of the terms for the chunk factor initializations. ref- https://github.com/blevesearch/zap/blob/dbfed402cb7601951953f0b6a65ac3592be321d6/merge.go#L315-L334

And this would have an extra IO/disk reach overheads which would be slowing us down in these merge code paths.

But no alternate/workaround idea seems feasible for this now as this looks like an inevitable step.

2. With the custom chunk factor introduction, we had removed some older merge path optimizations by copying the freq/norm/location bytes. ref - mergeTermFreqNormLocsByCopying

ref - https://github.com/blevesearch/zap/blob/27199d5dfc5d0bc33174bdd959442cb1223c48df/merge.go#L323

And this certainly looks like a useful optimization that can be resurrected as the custom chunk size/factor shouldn't affect these bytes copying operations during the merge.

 

Tried toy build which brings back this optimization and that certainly improved the performance of tests.

Example,

Wildcard =>   89 (master) => 117 (toy)

http://perf.jenkins.couchbase.com/job/atlas_dev/457/console

Fuzzy -1 ==>   671 (master) => 703 (toy)

http://perf.jenkins.couchbase.com/job/atlas_dev/460/console

Fuzzy -2 ==>   47 (master) => 61 (toy)

http://perf.jenkins.couchbase.com/job/atlas_dev/458/console

 

Interestingly, the performance hit/impact is perceivable only on vellum automaton-based queries like wildcard, fuzzy, prefix, etc. 

ie, CPU/IO/Memory load combinations of automaton based [*higher* *latency*] queries are more impacted by this slowness in the merge.

In other words, the slower merge operations regressed to result in more memory segments only in the presence of these automaton based queries(potentially attributed to the very nature of those automaton based queries), which resulted in lower query throughputs.__

**Comment by mschoch** [comment]
You've said you see perf improvement with this toybuild, however I didn't see anywhere if you said that all 3 nodes looked the same (normal) again?

**Comment by sreekanth-cb** [comment]
The factors playing here could be anything from a combination of multiple things like the nature of data causing a bigger automaton instances build-up that interferes with the slowness and other overheads of a wildcard query ...

In short, I don't have an answer yet.

**Comment by mschoch** [comment]
I think we should look closer again at how the mutations are being made.  It seems like it could indirectly affect the `fieldsSame` evaluation, which then connects to this behavior...

**Comment by sreekanth-cb** [comment]
Severity of the number of memory segment spiking came down. But the regression isn't fully fixed as the merge ops are still slower compared to zap11 days.

ref - http://cbmonitor.sc.couchbase.com/reports/html/?snapshot=fts_700-98674602_run_test_199d&snapshot=fts_700-3583_run_test_47fb

Green is the toy build one.

**Comment by sreekanth-cb** [comment]
> I think we should look closer again at how the mutations are being made. It seems like it could indirectly affect the `fieldsSame` evaluation, which then connects to this behavior...

fieldSame is getting into play as we have a particular document schema with static fields in our tests. And they keep updating a given field in the document.

But the perf issue is still open, we couldn't fully resurrect the performance back early days due to overall slowness in the merge paths.

**Comment by mschoch** [comment]
OK, it seems to me this has done nothing to fix the unusual behavior on one of the nodes.  That doesn't mean it isn't good improvement, it just seems unrelated to that problem.

**Comment by sreekanth-cb** [comment]
> OK, it seems to me this has done nothing to fix the unusual behavior on one of the nodes. That doesn't mean it isn't good improvement, it just seems unrelated to that problem.

I don't follow this.

Why do you think the fix has done nothing to fix the problem on the erring node when the number of memory segments came down from the order of 1500 range to <500 range?

**Comment by mschoch** [comment]
Oh, simply if I look at the num_files_on_disk, one node (40) is behaviing completely differently than the other 2.  I would like to understand why this is.

**Comment by sreekanth-cb** [comment]
Ah.. i see..

Let me try to explain this.

The persister cleans up the older/stale data whenever it gets leeway time during it's 
persist cum merge work loop.
ref - https://github.com/blevesearch/bleve/blob/master/index/scorch/persister.go#L194
https://github.com/blevesearch/bleve/blob/master/index/scorch/persister.go#L212

And it works perfectly as long it runs ahead of merger which is the usual case. 
Why?  since it gets ample opportunities of removeOldData() windows due to self-induced
slowness for the merger catch up too. (in addition to the first examples above)
ref - https://github.com/blevesearch/bleve/blob/master/index/scorch/persister.go#L279

In this particular problematic case, persister was falling behind the merger due to the
slow in memory merging. (this was evident from the fts epoch number stats)

And due to that - it always has catchup persist work loops to finish (as both introducer 
and the merger are ahead of it). This reduces the removeOldData() window or opportunity
for stale file cleanups.

**Comment by mschoch** [comment]
I'm still confused because you are using the past tense, as if the problem was fixed.  However, I still see node 40 behaving different from the other nodes.  You have explained why a node might exhibit this pattern, but not why only one node in the cluster does given this testcase.  My understanding is we are making same type of mutations to data that should be evenly distributed across all nodes.  Therefore, I expect all nodes to exhibit similar behavior.  Do you disagree?

**Comment by sreekanth-cb** [comment]
Let us sync on this during the weekly sync.

I tried to summarise a few obvious problems/causes of merge slowness here -https://github.com/blevesearch/zap/pull/44#issuecomment-720466507

And it's not entirely fixed and hence the problem is NOT expected to be completely solved too.

**Comment by mschoch** [comment]
Again, I understand that there is still some slowness, and that you believe that if were to fix that, that the behaviors would end up looking more the same (due to merger/persister interactions).  And I don't doubt this is true...   But it's still very unusual (and without any explanation) that it affects one node so much differently than the others.

**Comment by sreekanth-cb** [comment]
yes. that is right.

My half-cent theory/wild guess here is that due to the nature of data/mutations, many of these queries result in creating gigantic(holding many candidate terms) FST automatons that are interfering significantly with the other cpu/io impacting merge-operations or vice versa.

And if we look at the regressions, they are primarily on automaton based queries (fuzzy/wildcard/prefix), and the order of impacts are potentially higher with the bigger automaton overheads.. (fuzzy2 and wildcards are more affected than fuzzy1 and prefix) (thinking aloud) ...

**Comment by sreekanth-cb** [comment]
hey @mschoch,  I don't see any problems in proceeding with this change for the beta release. Any thoughts/feedback on it?

**Comment by mschoch** [comment]
The only problem is that it now takes about half a day to make a release.

**Review by mschoch — APPROVED** [review]
